### PR TITLE
Transition to Tailwind CSS

### PR DESCRIPTION
**Pull request by purfectliterature**
~~Note: This PR has [`phillmont/new-edit-assessment`](https://github.com/Coursemology/coursemology2/tree/phillmont/new-edit-assessment) (#5034) as a base branch.~~

---

<h1 align="center">
Transition to 
  <a href="https://tailwindcss.com/#gh-light-mode-only" target="_blank">
    Tailwind CSS
  </a>
  <a href="https://tailwindcss.com/#gh-dark-mode-only" target="_blank">
    Tailwind CSS
  </a>
</h1>

Tailwind CSS is a [utility-first CSS framework](https://tailwindcss.com/docs/utility-first) for rapidly building custom user interfaces. It offers a different paradigm of traditionally writing 'semantic' CSS, allowing for definitive stylesheet consistency, ease of configuration, and better developer experience. Read [how Adam Wathan, the creator of Tailwind CSS, convinced the world that utility-first CSS is a thing](https://adamwathan.me/css-utility-classes-and-separation-of-concerns/), and brought Tailwind CSS to [one of the best and most popular CSS frameworks](https://dev.to/theme_selection/best-css-frameworks-in-2020-1jjh).

## TL;DR
1. Moving forward, we are using Tailwind's CSS utilities for any styling. Ideally, MUI is only used to provide components.
2. Generally, there mustn't be `sx` or `style` props/attributes being used.
3. Inline styles, raw CSS, Sass stylesheets, [`styled-components`](https://mui.com/material-ui/guides/interoperability/#styled-components), or [Emotion](https://mui.com/material-ui/guides/interoperability/#emotion) are henceforth **illegal**.
4. As much as possible, refrain from using arbitrary values, [even if Tailwind supports it](https://tailwindcss.com/docs/adding-custom-styles#using-arbitrary-values).  
If you really have to, refrain from using absolute values. Use `rem`, not `em`.
5. If you are deep styling MUI components, i.e., modifying very specific parts of MUI components, use their official APIs, e.g., `componentProps`, `InputLabelProps` , `InputComponentProps`, etc., to provide Tailwind's `className`s. [This is MUI's official recommendation, anyway](https://mui.com/material-ui/guides/interoperability/#deeper-elements-5).
Unless really necessary, do NOT directly target MUI's class names, e.g., `.MuiDialog-root`, because ideally, these are MUI's internal implementation details that might change in their next versions.
6. [See an example of the New/Edit Assessment form's styles rewritten in Tailwind here](https://github.com/Coursemology/coursemology2/pull/5049/commits/5b8b71826584a279bb1b084bfe6d13b501a7cf8a).

## Things to note when developing with Tailwind for now
The Tailwind's stock utilities are provided in [`index.css`](https://github.com/Coursemology/coursemology2/pull/5049/files#diff-0759b90d78dade98c39a4fe72fed26d9f87aa20558084707bc4115d9a996bc05). This is also where you will put your [custom utilities](https://tailwindcss.com/docs/adding-custom-styles#adding-custom-utilities), if *really* needed. It must be loaded by the root React instance, ideally the one who calls `ReactDOM.render`. Since our SPA is upcoming, `index.css` is loaded by [`client/app/index.js`](https://github.com/Coursemology/coursemology2/pull/5049/files#diff-e7ee0a199a33c7454ed3a44e8279dd28a9927e73aa7901c25748c4a5e93ff5cf).

The default React `#root` element is self-assigned in [`course.html.slim`](https://github.com/Coursemology/coursemology2/pull/5049/files#diff-f479f0dbc7d66be0ec2e73c6a77478a1b40e8db1e7d3f24b87cc3b631e991638). If other modules need Tailwind utilities, append the `#root` ID to the parent HTML element.

If your Tailwind utilities do not work on any [`Portal`](https://mui.com/material-ui/react-portal/) elements, set their `container` to the `rootElement` ([like `Dialog`, `Popover`, and `Popper` here](https://github.com/Coursemology/coursemology2/pull/5049/files#:~:text=...adaptedTheme%2C-,components%3A%20%7B,-MuiDialog%3A%20%7B)) so they are mounted there instead of `document.body`.

## Some metric, if you care
As of this PR, with only some components' styles converted to Tailwind's utilities ([see Files changed](https://github.com/Coursemology/coursemology2/pull/5049/files)), their bundle sizes reduced by 3-5%, analyzed with [Bateman's Webpack Visualizer](https://chrisbateman.github.io/webpack-visualizer/). As more styles are moved to Tailwind, our bundle sizes should reduce, but only by a tad bit.

Notably, **performance or bundle size is not the point of this transition**. It is all about the developers' experience and writing stylesheets that are scalable and do not grow (sharply) as our React code grows.

### CODE REVIEW

**Comment by ekowidianto** [comment]
@purfectliterature Please rebase thanks!

**Comment by purfectliterature** [comment]
I found that `leaderboard_viewing_spec` tests for clicking the `#achievement-tab` button that only appears in screen sizes smaller than the `lg` breakpoint. I suppose because I matched our MUI breakpoints to Tailwind's breakpoints, `lg` reduced from 1200px (MUI default) to 1024px (Tailwind default), then the `#achievement-tab` button disappears because probably the test suite's chromedriver has screen size >1024px.

I committed a fix 732883db42e0a5eeba2a8454b0307488820d1446 that allows `leaderboard_viewing_spec` to test properly regardless of screen size. We should probably take note to write E2E tests that are media-agnostic.

**Review by ekowidianto — COMMENTED** [review]
Great job @purfectliterature!! Thanks for introducing this framework. I like it just because TLDR - ‘Tailwind is a constrained inline style that allows responsive design and hover, focus and other states styling’.

A few things to note:
1. Do we want to consider using this plugin to optimize prod build? Probably this wont help much now but I guess this is good for future development. https://tailwindcss.com/docs/optimizing-for-production
2. Could you help adding some docs on the usage for react styling in ‘Readme.md’ in the ‘client’ folder? We will need to start populating this file with more convention, guideline and usage info for our project (eg redux, intl, router, testing, etc).
3. Please add linter/eslint.
4. In the future, should we use tailwind theme or preset to centralize styling, theme ,color ,etc control and how/if do we reconcile with MUI?

File: client/app/lib/components/IconRadio.tsx [hunk 1]
```diff
@@ -9,58 +9,40 @@ interface IconRadioProps {
...
-  return (
-    <label
-      style={{
-        ...styles.container,
-        alignItems: props.description ? 'flex-start' : 'center',
-      }}
-    >
-      <Radio value={props.value} sx={styles.radio} disabled={props.disabled} />
+    {props.icon
+      ? createElement(props.icon, {
```
Comment (ekowidianto) [inline client/app/lib/components/IconRadio.tsx:21]: Change to &&

File: client/app/lib/components/IconRadio.tsx [hunk 2]
```diff
@@ -9,58 +9,40 @@ interface IconRadioProps {
...
+      : null}
 
-      {props.icon
-        ? createElement(props.icon, {
-            fontSize: 'large',
-            sx: styles.icon,
-            color: props.disabled ? 'disabled' : undefined,
-          })
-        : null}
+    <div>
```
Comment (ekowidianto) [inline client/app/lib/components/IconRadio.tsx:28]: Replace div?

File: client/app/lib/components/TextField.tsx [hunk 3]
```diff
@@ -18,7 +18,7 @@ const TextField = forwardRef<HTMLDivElement, TextFieldProps>(
     const [showPassword, setShowPassword] = useState(false);
 
     return (
-      <>
+      <div>
```
Comment (ekowidianto) [inline client/app/lib/components/TextField.tsx:21]: Just checking if we need a div here?

File: client/app/lib/components/ProviderWrapper.jsx [hunk 4]
```diff
@@ -24,13 +32,26 @@ const propTypes = {
...
   palette,
   // https://material-ui.com/customization/themes/#typography---html-font-size
   // https://material-ui.com/style/typography/#migration-to-typography-v2
   typography: {
     htmlFontSize: 10,
   },
+  breakpoints: {
+    values: {
+      xs: 0,
+      sm: pxInInt(tailwindConfig.theme.screens.sm),
```
Comment (ekowidianto) [inline client/app/lib/components/ProviderWrapper.jsx:49]: Besides these, are there any other theme settings to be reconciled between mui and tailwind?

File: client/app/lib/components/ProviderWrapper.jsx [hunk 5]
```diff
@@ -97,11 +118,21 @@ const themeSettings = {
...
 const ProviderWrapper = ({ store, persistor, children }) => {
   const localeWithoutRegionCode = i18nLocale.toLowerCase().split(/[_-]+/)[0];
 
+  // TODO: Replace with React's createRoot once true SPA is ready
+  const rootElement = document.getElementById('root');
+
+  const themeV5 = createTheme({
+    ...adaptedTheme,
+    components: {
+      MuiDialog: { defaultProps: { container: rootElement } },
```
Comment (ekowidianto) [inline client/app/lib/components/ProviderWrapper.jsx:130]: Could I check what would these be for?

File: client/app/lib/components/IconRadio.tsx (hunk 2 above)
Comment (purfectliterature) [reply client/app/lib/components/IconRadio.tsx:28]: To? The `div` is required here to arrange the `IconRadio`'s title and description vertically because the entire `label` element is a row flexbox.

File: client/app/lib/components/TextField.tsx (hunk 3 above)
Comment (purfectliterature) [reply client/app/lib/components/TextField.tsx:21]: Yes. There is a description `Typography` below the `TextField` component. Containing these two elements in a `div` ensures that the spacing between them will not be affected by the clients who `import` it.

File: client/tailwind.config.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+/** @type {import('tailwindcss').Config} */
+module.exports = {
+  content: ['./app/**/*.{js,jsx,ts,tsx}'],
+  theme: {
+    extend: {},
+  },
+  corePlugins: {
+    preflight: false,
```
[the remark this answers is not in the document]
Comment (purfectliterature) [reply client/tailwind.config.js:8]: I don't think so. [MUI's guide on Tailwind style library inoperability](https://mui.com/material-ui/guides/interoperability/#tailwind-css) asks us to explicitly remove Tailwind's preflight 'so it can use the MUI's preflight instead'. Using Tailwind's preflight will override MUI's style, e.g., our colours, buttons, font families, etc., that we have set in `Theme`.

Problem is, I'm not even sure if our codebase has MUI's preflight applied. MUI's preflight is the [`<CssBaseline>`](https://mui.com/material-ui/react-css-baseline/) component, but I don't see it anywhere, nor do I know if it is applied automatically. Preflights are usually automatic in CSS frameworks, and based on the quote above, it sounds like MUI's preflight is automatic.

[the remark this answers is not in the document]
Comment (ekowidianto) [reply client/tailwind.config.js:8]: Thanks for clarifying!

File: client/app/lib/components/IconRadio.tsx (hunk 2 above)
Comment (ekowidianto) [reply client/app/lib/components/IconRadio.tsx:28]: As discussed offline, we dont use Box component for simple div thanks!

File: client/app/lib/components/TextField.tsx (hunk 3 above)
Comment (ekowidianto) [reply client/app/lib/components/TextField.tsx:21]: Same as above

**Comment by purfectliterature** [comment]
> 1. Do we want to consider using this plugin to optimize prod build? Probably this wont help much now but I guess this is good for future development. https://tailwindcss.com/docs/optimizing-for-production

I have since set up `cssnano` as of https://github.com/Coursemology/coursemology2/pull/5049/commits/18681ec0ccc1f44d8cd0b679cd68b1d02b0e2198.

> 2. Could you help adding some docs on the usage for react styling in ‘Readme.md’ in the ‘client’ folder? We will need to start populating this file with more convention, guideline and usage info for our project (eg redux, intl, router, testing, etc).

I will add the documentations after I spend some time with them when I migrate the Course Settings page. I have added an issue #5067 for me to track.

> 3. Please add linter/eslint.

Tailwind has linting built into their [Tailwind CSS IntelliSense extension for Visual Studio Code](https://tailwindcss.com/blog/introducing-linting-for-tailwindcss-intellisense). This extension includes both linting and IntelliSense for their utilities, as well as the raw CSS outputs to see the actual `rem` values or actual colours.

> 4. In the future, should we use tailwind theme or preset to centralize styling, theme ,color ,etc control and how/if do we reconcile with MUI?

Yes, we should centralise everything in Tailwind, more specifically `tailwind.config.js`. This will make it breezy when it comes to custom theming in the future. This includes breakpoints (done), colours, text systems, spacing, border styles, transitions, animations, and/or others that I cannot think on top of my head right now.

File: client/app/lib/components/ProviderWrapper.jsx (hunk 4 above)
Comment (purfectliterature) [reply client/app/lib/components/ProviderWrapper.jsx:49]: See my comment below for number 4.

**Comment by ekowidianto** [comment]
> > 1. Do we want to consider using this plugin to optimize prod build? Probably this wont help much now but I guess this is good for future development. https://tailwindcss.com/docs/optimizing-for-production
> 
> I have since set up `cssnano` as of [8780a4c](https://github.com/Coursemology/coursemology2/commit/8780a4c0073c950353f6f46653eb7e43c166edca).
> 
> > 2. Could you help adding some docs on the usage for react styling in ‘Readme.md’ in the ‘client’ folder? We will need to start populating this file with more convention, guideline and usage info for our project (eg redux, intl, router, testing, etc).
> 
> I will add the documentations after I spend some time with them when I migrate the Course Settings page. I have added an issue #5067 for me to track.
> 
> > 3. Please add linter/eslint.
> 
> Tailwind has linting built into their [Tailwind CSS IntelliSense extension for Visual Studio Code](https://tailwindcss.com/blog/introducing-linting-for-tailwindcss-intellisense). This extension includes both linting and IntelliSense for their utilities, as well as the raw CSS outputs to see the actual `rem` values or actual colours.
> 
> > 4. In the future, should we use tailwind theme or preset to centralize styling, theme ,color ,etc control and how/if do we reconcile with MUI?
> 
> Yes, we should centralise everything in Tailwind, more specifically `tailwind.config.js`. This will make it breezy when it comes to custom theming in the future. This includes breakpoints (done), colours, text systems, spacing, border styles, transitions, animations, and/or others that I cannot think on top of my head right now.
> 
> Text systems and colours can be a bit iffy, because we'd probably still want to be able to access these via MUI's props, e.g., `<Typography variant="body1" ...>` or `<Icon color="text.secondary" ...>`. We probably will need to formalise our colour palette to `tailwind.config.js` and retrieve the colours to MUI `Theme`. Otherwise, we can only do `<Icon className="text-gray-200" ...>` for colour, or (a) `<Typography className="text-sm" ...>` or (b) create our own `@layer utilities { .text-body-1 { ... } }` in `index.css` and do `<Typography className="text-body-1" ...>`, for text systems.
> 
> I did not set this up in this PR because (a) it could affect font/colour styles in many, many pages, and (b) our transition to Tailwind is still under probation anyway. Let me fully test-drive Tailwind in my Course Settings migration first, then once everything is okay and there is no gotchas, I will set up the remaining bits and pieces of MUI x Tailwind; we can merge this PR in first.

Great thanks!

**Review by ekowidianto — APPROVED** [review]
